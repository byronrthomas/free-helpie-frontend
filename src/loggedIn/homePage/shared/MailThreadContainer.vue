<template>
  <div>
    <button 
      v-if="showMakeConnectionButton"
      class="btn btn-primary"
      id="makeConnection"
      @click="$emit('makeConnection')">
    Connect with user</button>
    <button
      v-if="showCancelConnectionButton"
      class="btn btn-primary"
      id="cancelConnection"
      @click="$emit('cancelConnection')">
    Cancel connection with user</button>
    <div v-if="!waitingForReply">
      <textarea 
        name="messageContents" 
        id="messageContents" 
        cols="30" rows="10" 
        v-model="newMailText"></textarea>
      <button 
        :disabled="!newMailText" 
        class="btn btn-primary" 
        id="submitButton" 
        @click="sendMail">Send</button>
    </div>
    <mail-item
      v-for="mail in mailItems"
      :key="mail.id"
      :mail="mail"
      :avatar="avatarImage(mail)"/>
  </div>
</template>

<script>
import MailItem from './MailItem.vue'

function compareDates (aDate, bDate) {
  return aDate.getTime() - bDate.getTime()
}

export default {
  props: {
    mailItems: Array,
    username: String,
    myAvatar: Object,
    otherAvatar: Object,
    connectOrCancelAllowed: {
      type: String,
      default () {
        return 'disabled'
      },
      validator (value) {
        return ['disabled', 'connectAllowed', 'cancelAllowed'].includes(value)
      }
    }
  },
  data () {
    const sortedMailItems = [...this.mailItems]
    sortedMailItems.sort((a, b) => compareDates(b.sent, a.sent))

    return {
      newMailText: '',
      sortedMailItems: sortedMailItems
    }
  },
  computed: {
    waitingForReply () {
      return this.sortedMailItems.length > 0 &&
        this.sentByCurrentUser(this.sortedMailItems[0])
    }
  },
  methods: {
    sentByCurrentUser (mail) {
      return this.username === mail.sender
    },
    avatarImage (mail) {
      return this.sentByCurrentUser(mail)
        ? this.myAvatar
        : this.otherAvatar
    },
    sendMail () {
      console.log('Getting a sendMail click')
      console.log(this.newMailText)
      if (this.newMailText) {
        console.log('Sending mail:')
        console.log(this.newMailText)
        this.$emit('sendMail', this.newMailText)
      }
    },
    showMakeConnectionButton () {
      return this.connectOrCancelAllowed === 'connectAllowed'
    },
    showCancelConnectionButton () {
      return this.connectOrCancelAllowed === 'cancelAllowed'
    }
  },
  components: {
    'mail-item': MailItem
  }
}
</script>

