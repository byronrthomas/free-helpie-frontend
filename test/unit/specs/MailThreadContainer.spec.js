import { mount } from 'vue-test-utils'

import MailThreadContainer from '@/loggedIn/homePage/shared/MailThreadContainer.vue'
import { fail } from 'assert';

function mailWithDate (sentDate) {
  return {
    id: 0,
    sent: sentDate}
}

function makeOptionsForConnectCancel (connectCancel) {
  return {
    propsData: {
      mailItems: [],
      myAvatar: {altText: 'You'},
      otherAvatar: {altText: 'Them'},
      username: TEST_USERNAME,
      connectOrCancelAllowed: connectCancel}}
}

const TEST_USERNAME = 'TestUser'
function makeOptions (mails) {
  return {
    propsData: {
      mailItems: mails,
      myAvatar: {altText: 'You'},
      otherAvatar: {altText: 'Them'},
      username: TEST_USERNAME}}
}

let nextDate = new Date(2014, 10, 10)
function makeAMailSentBy (user) {
  const nextTime = nextDate.getTime() + 1000
  nextDate = new Date()
  nextDate.setTime(nextTime)
  return {
    id: 1,
    replyToMailId: -1,
    sent: nextDate,
    sender: user
  }
}

const NO_MAILS = makeOptions([])

describe('Mail thread container', () => {
  it('should sort mail items in reverse sent order', () => {
    const inputMails = [
      new Date(2010, 1, 1, 11, 11, 5),
      new Date(2011, 1, 1, 11, 11, 5),
      new Date(2010, 1, 1, 11, 10, 5),
      new Date(2010, 2, 1, 11, 11, 5),
    ].map(mailWithDate)

    const expectedOrder = 
      [inputMails[1], inputMails[3], inputMails[0], inputMails[2]]

    const onTest = mount(MailThreadContainer, makeOptions(inputMails))
    expect(onTest.vm.sortedMailItems).toEqual(expectedOrder)
  })

  it('should return !sentByCurrentUser if the sender of the mail is not the current user', () => {
    const onTest = mount(MailThreadContainer, NO_MAILS)
    expect(onTest.vm.sentByCurrentUser({sender: 'SomebodyElse'})).toBeFalsy()
  })

  it('should return sentByCurrentUser if the sender of the mail is the current user', () => {
    const onTest = mount(MailThreadContainer, NO_MAILS)
    expect(onTest.vm.sentByCurrentUser({sender: onTest.vm.username})).toBeTruthy()
  })

  it('should emit sendMail with the new mail text when the button is clicked', () => {
    const emailText = 'Email text'
    const onTest = mount(MailThreadContainer, NO_MAILS)
    onTest.setData({...onTest.vm.$data, newMailText: emailText})
    onTest.find('#submitButton').trigger('click')

    expect(onTest.emitted().sendMail).toBeTruthy()
    expect(onTest.emitted().sendMail.length).toBe(1)
    expect(onTest.emitted().sendMail[0]).toEqual([emailText])
  })

  it('should emit makeConnection when the button is clicked', () => {
    const onTest = mount(MailThreadContainer, makeOptionsForConnectCancel('connectAllowed'))
    onTest.find('#makeConnection').trigger('click')

    expect(onTest.emitted().makeConnection).toBeTruthy()
    expect(onTest.emitted().makeConnection.length).toBe(1)
    expect(onTest.emitted().makeConnection[0]).toEqual([])
  })

  it('should emit cancelConnection when the button is clicked', () => {
    const onTest = mount(MailThreadContainer, makeOptionsForConnectCancel('cancelAllowed'))
    onTest.find('#cancelConnection').trigger('click')

    expect(onTest.emitted().cancelConnection).toBeTruthy()
    expect(onTest.emitted().cancelConnection.length).toBe(1)
    expect(onTest.emitted().cancelConnection[0]).toEqual([])
  })

  const allowedText = allowed => allowed ? 'allowed' : 'not allowed'
  const aMail = mailWithDate(new Date(2017, 11, 22))
  for (const hasMessages of [false, true]) {
    const messagesText = hasMessages ? 'messages' : 'no messages'
    for (const connectOrCancelState of ['disabled', 'connectAllowed', 'cancelAllowed']) {
      it(`should be rendered correctly when there are ${messagesText} and the cancelConnect state is ${connectOrCancelState}`, () => {
        const msgs = hasMessages ? [{...aMail, sender: 'AnotherUser'}] : []
        const options = {
          propsData: {
            mailItems: msgs,
            myAvatar: {altText: 'You'},
            otherAvatar: {altText: 'Them'},
            username: TEST_USERNAME},
            connectOrCancelAllowed: connectOrCancelState}
        const onTest = mount(MailThreadContainer, options)
        expect(onTest.element).toMatchSnapshot()
      })
    }
  }
})