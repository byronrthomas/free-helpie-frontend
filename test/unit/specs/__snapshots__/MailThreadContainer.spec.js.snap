// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mail thread container should be rendered correctly when there are messages and the cancelConnect state is cancelAllowed 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-xs-12"
    >
      <div
        class="floatable-bordered"
      >
        <h2
          style="text-align: center;"
        >
          
        </h2>
         
        <br />
      </div>
       
      <div
        class="floatable-bordered floatable-bordered-bottom"
      >
        <br />
         
        <div
          class="row"
        >
          <div
            class="col-xs-10 col-xs-offset-1"
          >
            <div
              class="panel panel-default"
            >
              <div
                class="panel-heading"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-sm-3"
                  >
                    <strong>
                       Them 
                    </strong>
                  </div>
                   
                  <div
                    class="col-sm-3 col-sm-offset-6"
                  >
                    <p
                      style="float: right;"
                    >
                      2014-11-10 00:00:01
                    </p>
                  </div>
                </div>
              </div>
               
              <div
                class="panel-body"
              >
                
    
  
              </div>
            </div>
             
            <div
              class="panel panel-primary"
            >
              <div
                class="panel-heading"
              >
                <h3
                  class="panel-title"
                >
                  Send a message
                </h3>
              </div>
               
              <div
                class="panel-body"
              >
                <textarea
                  class="fullsizeTextArea"
                  id="messageContents"
                  name="messageContents"
                />
                 
                <button
                  class="btn btn-primary"
                  disabled="disabled"
                  id="submitButton"
                >
                  Send
                </button>
              </div>
            </div>
             
            <!---->
             
            <!---->
          </div>
        </div>
      </div>
       
      <br />
    </div>
  </div>
</div>
`;

exports[`Mail thread container should be rendered correctly when there are messages and the cancelConnect state is connectAllowed 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-xs-12"
    >
      <div
        class="floatable-bordered"
      >
        <h2
          style="text-align: center;"
        >
          
        </h2>
         
        <br />
      </div>
       
      <div
        class="floatable-bordered floatable-bordered-bottom"
      >
        <br />
         
        <div
          class="row"
        >
          <div
            class="col-xs-10 col-xs-offset-1"
          >
            <div
              class="panel panel-default"
            >
              <div
                class="panel-heading"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-sm-3"
                  >
                    <strong>
                       Them 
                    </strong>
                  </div>
                   
                  <div
                    class="col-sm-3 col-sm-offset-6"
                  >
                    <p
                      style="float: right;"
                    >
                      2014-11-10 00:00:01
                    </p>
                  </div>
                </div>
              </div>
               
              <div
                class="panel-body"
              >
                
    
  
              </div>
            </div>
             
            <div
              class="panel panel-primary"
            >
              <div
                class="panel-heading"
              >
                <h3
                  class="panel-title"
                >
                  Send a message
                </h3>
              </div>
               
              <div
                class="panel-body"
              >
                <textarea
                  class="fullsizeTextArea"
                  id="messageContents"
                  name="messageContents"
                />
                 
                <button
                  class="btn btn-primary"
                  disabled="disabled"
                  id="submitButton"
                >
                  Send
                </button>
              </div>
            </div>
             
            <!---->
             
            <!---->
          </div>
        </div>
      </div>
       
      <br />
    </div>
  </div>
</div>
`;

exports[`Mail thread container should be rendered correctly when there are messages and the cancelConnect state is disabled 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-xs-12"
    >
      <div
        class="floatable-bordered"
      >
        <h2
          style="text-align: center;"
        >
          
        </h2>
         
        <br />
      </div>
       
      <div
        class="floatable-bordered floatable-bordered-bottom"
      >
        <br />
         
        <div
          class="row"
        >
          <div
            class="col-xs-10 col-xs-offset-1"
          >
            <div
              class="panel panel-default"
            >
              <div
                class="panel-heading"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-sm-3"
                  >
                    <strong>
                       Them 
                    </strong>
                  </div>
                   
                  <div
                    class="col-sm-3 col-sm-offset-6"
                  >
                    <p
                      style="float: right;"
                    >
                      2014-11-10 00:00:01
                    </p>
                  </div>
                </div>
              </div>
               
              <div
                class="panel-body"
              >
                
    
  
              </div>
            </div>
             
            <div
              class="panel panel-primary"
            >
              <div
                class="panel-heading"
              >
                <h3
                  class="panel-title"
                >
                  Send a message
                </h3>
              </div>
               
              <div
                class="panel-body"
              >
                <textarea
                  class="fullsizeTextArea"
                  id="messageContents"
                  name="messageContents"
                />
                 
                <button
                  class="btn btn-primary"
                  disabled="disabled"
                  id="submitButton"
                >
                  Send
                </button>
              </div>
            </div>
             
            <!---->
             
            <!---->
          </div>
        </div>
      </div>
       
      <br />
    </div>
  </div>
</div>
`;

exports[`Mail thread container should be rendered correctly when there are no messages and the cancelConnect state is cancelAllowed 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-xs-12"
    >
      <div
        class="floatable-bordered"
      >
        <h2
          style="text-align: center;"
        >
          
        </h2>
         
        <br />
      </div>
       
      <div
        class="floatable-bordered floatable-bordered-bottom"
      >
        <br />
         
        <div
          class="row"
        >
          <div
            class="col-xs-10 col-xs-offset-1"
          >
             
            <div
              class="panel panel-primary"
            >
              <div
                class="panel-heading"
              >
                <h3
                  class="panel-title"
                >
                  Send a message
                </h3>
              </div>
               
              <div
                class="panel-body"
              >
                <textarea
                  class="fullsizeTextArea"
                  id="messageContents"
                  name="messageContents"
                />
                 
                <button
                  class="btn btn-primary"
                  disabled="disabled"
                  id="submitButton"
                >
                  Send
                </button>
              </div>
            </div>
             
            <!---->
             
            <!---->
          </div>
        </div>
      </div>
       
      <br />
    </div>
  </div>
</div>
`;

exports[`Mail thread container should be rendered correctly when there are no messages and the cancelConnect state is connectAllowed 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-xs-12"
    >
      <div
        class="floatable-bordered"
      >
        <h2
          style="text-align: center;"
        >
          
        </h2>
         
        <br />
      </div>
       
      <div
        class="floatable-bordered floatable-bordered-bottom"
      >
        <br />
         
        <div
          class="row"
        >
          <div
            class="col-xs-10 col-xs-offset-1"
          >
             
            <div
              class="panel panel-primary"
            >
              <div
                class="panel-heading"
              >
                <h3
                  class="panel-title"
                >
                  Send a message
                </h3>
              </div>
               
              <div
                class="panel-body"
              >
                <textarea
                  class="fullsizeTextArea"
                  id="messageContents"
                  name="messageContents"
                />
                 
                <button
                  class="btn btn-primary"
                  disabled="disabled"
                  id="submitButton"
                >
                  Send
                </button>
              </div>
            </div>
             
            <!---->
             
            <!---->
          </div>
        </div>
      </div>
       
      <br />
    </div>
  </div>
</div>
`;

exports[`Mail thread container should be rendered correctly when there are no messages and the cancelConnect state is disabled 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-xs-12"
    >
      <div
        class="floatable-bordered"
      >
        <h2
          style="text-align: center;"
        >
          
        </h2>
         
        <br />
      </div>
       
      <div
        class="floatable-bordered floatable-bordered-bottom"
      >
        <br />
         
        <div
          class="row"
        >
          <div
            class="col-xs-10 col-xs-offset-1"
          >
             
            <div
              class="panel panel-primary"
            >
              <div
                class="panel-heading"
              >
                <h3
                  class="panel-title"
                >
                  Send a message
                </h3>
              </div>
               
              <div
                class="panel-body"
              >
                <textarea
                  class="fullsizeTextArea"
                  id="messageContents"
                  name="messageContents"
                />
                 
                <button
                  class="btn btn-primary"
                  disabled="disabled"
                  id="submitButton"
                >
                  Send
                </button>
              </div>
            </div>
             
            <!---->
             
            <!---->
          </div>
        </div>
      </div>
       
      <br />
    </div>
  </div>
</div>
`;
